// typography/dist/index.js
'use strict';

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _compassVerticalRhythm = require('compass-vertical-rhythm');

var _compassVerticalRhythm2 = _interopRequireDefault(_compassVerticalRhythm);

var _modularscale = require('modularscale');

var _modularscale2 = _interopRequireDefault(_modularscale);

var _createStyles = require('./utils/createStyles');

var _createStyles2 = _interopRequireDefault(_createStyles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createStylesString = function createStylesString(options) {
  var vr = (0, _compassVerticalRhythm2.default)(options);
  var styles = (0, _createStyles2.default)(vr, options);

  return styles;
};

var typography = function typography(opts) {
  var defaults = {
    baseFontSize: '18px',
    baseLineHeight: '28.5px',
    modularScales: [{
      scale: 'octave'
    }, {
      scale: 'golden',
      maxWidth: '768px'
    }],
    googleFonts: [],
    headerFontFamily: ['-apple-system', 'BlinkMacSystemFont', 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', 'sans-serif'],
    bodyFontFamily: ['georgia', 'serif'],
    headerGray: 20,
    headerGrayHue: 0,
    bodyGray: 20,
    bodyGrayHue: 0,
    headerWeight: 900,
    bodyWeight: 'normal',
    boldWeight: 'bold',
    includeNormalize: true,
    fontFaces: []
  };

  var options = (0, _objectAssign2.default)({}, defaults, opts);

  var vr = (0, _compassVerticalRhythm2.default)(options);

  return {
    options: options,
    rhythm: vr.rhythm,
    createStyles: function createStyles() {
      return createStylesString(options);
    },

    fontSizeToPx: vr.adjustFontSizeTo,
    fontSizeToMS: function fontSizeToMS(scaler) {
      // This doesn't pick the right scale if a theme has more than one scale.
      // Perhaps add optional parameter for a width and it'll get the scale
      // for this width. Tricky part is maxWidth could be set in non-pixels.
      var baseFont = options.baseFontSize.slice(0, -2);
      var newFontSize = (0, _modularscale2.default)(scaler, options.modularScales[0].scale) * baseFont + 'px';
      return vr.adjustFontSizeTo(newFontSize);
    },
    toString: function toString() {
      return this.createStyles();
    },
    injectStyles: function injectStyles() {
      if (typeof document !== 'undefined') {
        // Replace existing
        if (document.getElementById('typography.js')) {
          var styleNode = document.getElementById('typography.js');
          styleNode.innerHTML = createStylesString(options);
        } else {
          var node = document.createElement('style');
          node.id = 'typography.js';
          node.innerHTML = createStylesString(options);
          document.head.appendChild(node);
        }
      }
    }
  };
};

module.exports = typography;

/*
console.log(Typography({
  baseFontSize: '16px',
  includeNormalize: false,
}).toString())
*/
